networkx/algorithms/lowest_common_ancestors.py

```diff
@@ -72,34 +72,47 @@ def naive_all_pairs_lowest_common_ancestor(G, pairs=None):
     elif None in G:
         raise nx.NetworkXError("None is not a valid node.")
 
-    ancestor_cache = {}
-
     if pairs is None:
-
         pairs = combinations_with_replacement(G, 2)
-
-    for v, w in pairs:
-        if v not in ancestor_cache:
-            ancestor_cache[v] = nx.ancestors(G, v)
-            ancestor_cache[v].add(v)
-        if w not in ancestor_cache:
-            ancestor_cache[w] = nx.ancestors(G, w)
-            ancestor_cache[w].add(w)
-
-        common_ancestors = ancestor_cache[v] & ancestor_cache[w]
-
-        if common_ancestors:
-            common_ancestor = next(iter(common_ancestors))
-            while True:
-                successor = None
-                for lower_ancestor in G.successors(common_ancestor):
-                    if lower_ancestor in common_ancestors:
-                        successor = lower_ancestor
+    else:
+        # Convert iterator to iterable, if necessary, trim duplicates
+        pairs = set(pairs)
```

Comment:
Using a set here can/will break the order in which the pairs+lca are yielded. If order is important, `list` can be used instead.

Reply:
Good idea! Ultimately I think it's strange that the function guarantees *both* that it eagerly raises `NodeNotFound` exceptions for `pairs` *and* excepts generators for the `pairs` argument --- The former essentially guarantees that the generator either needs to be consumed and stored in some iterable under-the-hood (as is currently done, obviating the advantages of using a generator), or the generator needs to be cycled through twice.

However this is a discussion for another issue/PR! For now I'll focus on getting the behavior exactly the same with *as minimal* changes as possible.

Reply:
I understand the strangeness about both eagerly checking for nodes and accepting a generator. But it is more friendly for the user not to have to construct e.g. a list if they have an iterator. And they don't have to decide whether it should be a list/dict/set/tuple. At the same time it is more friendly to raise NodeNotFound eagerly. What is the downside? Perhaps that there are too many different possible inputs, which makes the code harder to read/follow.

It certainly means that we are not able to maintain the generator processing flow. We convert it to a dict/set in order to test the nodes.  If we were designing from scratch I would be tempted to make the input ignore pairs which contain nodes not in G. Then we can eliminate the NodeNotFound checking code...  

I think there is probably a good general principle here -- but I'm not sure what it is :)
As you say -- maybe that's for another PR/issue.